add GPU manual (#1408)
qdrant/landing_page: qdrant-landing/content/documentation/guides/running-with-GPU.md

---

> title: Running with GPU

Author: What if we add some other GPU feature later? This document should stay the same

> title: Running with GPU

Reviewer: ```suggestion
title: Running Qdrant with GPU Support
```

> title: Running with GPU

Author: Too long for sidebar

    # against the gpu device name.
    # Example: "nvidia"
    # Default: "" - all devices are accepted.
    device_filter: ""
    # List of explicit GPU devices to use.
    # If host has multiple GPUs, this option allows to select specific devices
    # by their index in the list of found devices.
    # If `device_filter` is set, indexes are applied after filtering.
    # By default, all devices are accepted.
    devices: null
    # How many parallel indexing processes are allowed to run.
    # Default: 1
    parallel_indexes: 1
    # Allow to use integrated GPUs.
    # Default: false
    allow_integrated: false
    # Allow to use emulated GPUs like LLVMpipe. Useful for CI.
    # Default: false
    allow_emulated: false
```

It is not recommended to change these options unless you are familiar with the Qdrant internals and the Vulkan API.

## Standalone

For standalone usage, you can build Qdrant with GPU support by running the following command:

```bash
cargo build --release --features gpu
```

To use GPUs, the device should support Vulkan API v1.3. That means Apple Silicon, Intel GPUs, CPU emulators, etc should work correctly.

## NVIDIA GPUs

### Prerequires

> To run Docker with NVidia support, your host should have an actual driver and [nvidia-container-toolkit](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/install-guide.html). Most default AI or CUDA images on Amazon/GCP/etc. are configured with nvidia container toolkit.

Reviewer: ```suggestion
To use Docker with NVIDIA GPU support, ensure the following are installed on your host:
- Latest NVIDIA drivers
- [nvidia-container-toolkit](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/install-guide.html)

Most AI or CUDA images on Amazon/GCP come pre-configured with the NVIDIA container toolkit.
```

### Docker images with NVidia GPU support

> Nvidia docker images have the same tag as Qdrant but with `gpu-nvidia` suffix, for example `qdrant/qdrant:v1.13.0-gpu-nvidia`.

Reviewer: ```suggestion
Docker images with NVIDIA GPU support use the tag suffix `gpu-nvidia`, e.g., `qdrant/qdrant:v1.13.0-gpu-nvidia`. These images include all necessary dependencies.
```

> This docker image contains all required dependencies to run Qdrant with GPU support.

Reviewer: ```suggestion
```

> To enable GPU for docker, you need an additional `--gpus=all` flag. To enable GPU for Qdrant you need to set enable flag from settings. Example of starting Qdrant with NVidia GPU support:

Reviewer: ```suggestion
To enable GPU support, use the `--gpus=all` flag with Docker settings. Example:
```

```bash
# `--gpus=all` flag says to Docker that we want to use GPUs.
# `-e QDRANT__GPU__indexing=1` flag says to QDrant that we want to use GPUs for indexing.
docker run \
	--rm \
	--gpus=all \
	-p 6333:6333 \
	-p 6334:6334 \
	-e QDRANT__GPU__indexing=1 \
	qdrant/qdrant:gpu-nvidia-latest
```

To ensure that the GPU was initialized correctly, you may check it in logs. First Qdrant prints all found GPU devices without filtering and then prints list of all created devices:

```
2025-01-13T11:58:29.124087Z  INFO gpu::instance: Foung GPU device: NVIDIA GeForce RTX 3090    
2025-01-13T11:58:29.124118Z  INFO gpu::instance: Foung GPU device: llvmpipe (LLVM 15.0.7, 256 bits)    
2025-01-13T11:58:29.124138Z  INFO gpu::device: Create GPU device NVIDIA GeForce RTX 3090    
```

> In this example logs you may see that 2 devices were found: RTX 3090 and llvmpipe (a CPU-emulated GPU which is included into docker image). And later you may see that only RTX was initialized.

Reviewer: ```suggestion
Here you can that two devices were found: RTX 3090 and llvmpipe (a CPU-emulated GPU which is included in the Docker image). Later, you will see that only RTX was initialized.
```

Author: Here you can `see`?

> That’s it. In a basic scenario, there is nothing to do more.

Reviewer: ```suggestion
This concludes the setup. Now, you can start using this Qdrant instance.
```

> ### Troubleshoots

Reviewer: ```suggestion
### Troubleshooting NVIDIA GPUs
```

> If your GPU is not detected in docker, check first your driver and `nvidia-container-toolkit` versions.

Reviewer: ```suggestion
If your GPU is not detected in Docker, make sure your driver and `nvidia-container-toolkit` are up-to-date. 

Verify Vulkan API visibility in the Docker container using:
```

> They both should be latest.

Reviewer: ```suggestion
```

> To check that Vulkan API is visible in a docker container, try to run this command:

Reviewer: ```suggestion
```

```
docker run --rm --gpus=all qdrant/qdrant:gpu-nvidia-latest vulkaninfo --summary
```

> You may find an error message with an explanation of why the NVidia device is not visible.

Reviewer: ```suggestion
The system may show you an error message explaining why the NVIDIA device is not visible.
```

> If your NVidia GPU is not visible in Docker, Docker image cannot use libGLX_nvidia.so.0 on your host. Example of an error message:

Reviewer: ```suggestion
Note that if your NVIDIA GPU is not visible in Docker, the Docker image cannot use libGLX_nvidia.so.0 on your host. Here is what an error message could look like:
```

```
ERROR: [Loader Message] Code 0 : loader_scanned_icd_add: Could not get `vkCreateInstance` via `vk_icdGetInstanceProcAddr` for ICD libGLX_nvidia.so.0
WARNING: [Loader Message] Code 0 : terminator_CreateInstance: Failed to CreateInstance in ICD 0. Skipping ICD.
```

> To fix it, try to change the config:

Reviewer: ```suggestion
To resolve errors, update your NVIDIA container runtime configuration:
```

```
sudo nano /etc/nvidia-container-runtime/config.toml
```

> Set `no-cgroups=false`, save config, and restart docker

Reviewer: ```suggestion
Set `no-cgroups=false`, save the configuration, and restart Docker:
```

```
sudo systemctl restart docker
```

## AMD GPUs

> AMD docker images have the same tag as qdant but with `gpu-amd` suffix, for example `qdrant/qdrant:v1.13.0-gpu-amd`.

Reviewer: ```suggestion
Docker images for AMD GPUs use the tag suffix `gpu-amd`, e.g., `qdrant/qdrant:v1.13.0-gpu-amd`. These images include all required dependencies.
```

> This docker image contains all required dependencies to run Qdrant with GPU support.

Reviewer: ```suggestion
```

> To enable GPU for docker, you need an additional `--device /dev/kfd --device /dev/dri` flags. To enable GPU for Qdrant you need to set enable flag. Example of starting Qdrant with AMD GPU support:

Reviewer: ```suggestion
To enable GPU for Docker, you need additional `--device /dev/kfd --device /dev/dri` flags. To enable GPU for Qdrant you need to set the enable flag. Here is an example:
```

```bash
# `--device /dev/kfd --device /dev/dri` flags say to Docker that we want to use GPUs.
# `-e QDRANT__GPU__indexing=1` flag says to QDrant that we want to use GPUs for indexing.
docker run \
	--rm \
	--device /dev/kfd --device /dev/dri \
	-e QDRANT__log_level=debug \
	-p 6333:6333 \
	-p 6334:6334 \
	-e QDRANT__GPU__indexing=1 \
	qdrant/qdrant:gpu-amd-latest
```

> To ensure that the GPU was initialized correctly, you may check it in logs:

Reviewer: ```suggestion
Check logs to confirm GPU initialization. Example log output:
```

```text
2025-01-10T11:56:55.926466Z  INFO gpu::instance: Foung GPU device: AMD Radeon Graphics (RADV GFX1103_R1)
2025-01-10T11:56:55.926485Z  INFO gpu::instance: Foung GPU device: llvmpipe (LLVM 17.0.6, 256 bits) 
2025-01-10T11:56:55.926504Z  INFO gpu::device: Create GPU device AMD Radeon Graphics (RADV GFX1103_R1)
```

> That’s it. In a basic scenario, there is nothing to do more.

Reviewer: ```suggestion
This concludes the setup. In a basic scenario, you won't need to configure anything else.
```

## Known limitations

> * Only Lnux x86_64 is supported. Windows, MacOS, ARM and other platforms are not supported yet.

Reviewer: ```suggestion
* **Platform Support:** Only Linux x86_64 is supported. Windows, macOS, ARM, and other platforms are not supported.
```

> * In current implementation, there is a limit of vector data, that can be processed by one GPU device in one indexing iteration - 16Gb.

Reviewer: ```suggestion
* **Memory Limits:** Each GPU can process up to 16GB of vector data per indexing iteration. 
```

> That means that you should not create segments, where either original vectors or (if you are usign quantization) quantized vectors are larger than 16Gb.

Reviewer: ```suggestion
Due to this limitation, you should not create segments where either original vectors OR quantized vectors are larger than 16GB.
```

    # Default: 512
    groups_count: 512
    # Filter for GPU devices by hardware name. Case insensitive.
    # Comma-separated list of substrings to match 
    # against the gpu device name.
    # Example: "nvidia"
    # Default: "" - all devices are accepted.
    device_filter: ""
    # List of explicit GPU devices to use.
    # If host has multiple GPUs, this option allows to select specific devices
    # by their index in the list of found devices.
    # If `device_filter` is set, indexes are applied after filtering.
    # By default, all devices are accepted.
    devices: null
    # How many parallel indexing processes are allowed to run.
    # Default: 1
    parallel_indexes: 1
    # Allow to use integrated GPUs.
    # Default: false
    allow_integrated: false
    # Allow to use emulated GPUs like LLVMpipe. Useful for CI.
    # Default: false
    allow_emulated: false
```

It is not recommended to change these options unless you are familiar with the Qdrant internals and the Vulkan API.

## Standalone GPU Support

For standalone usage, you can build Qdrant with GPU support by running the following command:

```bash
cargo build --release --features gpu
```

> Ensure your device supports Vulkan API v1.3. This enables compatibility with Apple Silicon, Intel GPUs, and CPU emulators.

Reviewer 2: I'm not sure if it makes sense to mention that the device should support some device features like `f16`, `u8` operations ect. All actual server grade gpus support everything. From all tested devices I got only old laptop nvidia MX150 GPU with does not support some features

That’s it. In a basic scenario, there is nothing to do more.

### Troubleshoots

If your GPU is not detected in docker, check first your driver and `nvidia-container-toolkit` versions.
They both should be latest.

To check that Vulkan API is visible in a docker container, try to run this command:

```
docker run --rm --gpus=all qdrant/qdrant:gpu-nvidia-latest vulkaninfo --summary
```

You may find an error message with an explanation of why the NVidia device is not visible.
If your NVidia GPU is not visible in Docker, Docker image cannot use libGLX_nvidia.so.0 on your host. Example of an error message:

```
ERROR: [Loader Message] Code 0 : loader_scanned_icd_add: Could not get `vkCreateInstance` via `vk_icdGetInstanceProcAddr` for ICD libGLX_nvidia.so.0
WARNING: [Loader Message] Code 0 : terminator_CreateInstance: Failed to CreateInstance in ICD 0. Skipping ICD.
```

To fix it, try to change the config:

```
sudo nano /etc/nvidia-container-runtime/config.toml
```

Set `no-cgroups=false`, save config, and restart docker

```
sudo systemctl restart docker
```

## AMD GPUs

> AMD docker images have the same tag as qdant but with `gpu-amd` suffix, for example `qdrant/qdrant:v1.13.0-gpu-amd`.

Reviewer 2: AMD host should have ROCm. Can you please all that user may install ROCm. Link https://rocm.docs.amd.com/projects/install-on-linux/en/latest/install/detailed-install.html. And you also may mention like in NVidia part that server images have already all installed soflware

This docker image contains all required dependencies to run Qdrant with GPU support.

To enable GPU for docker, you need an additional `--device /dev/kfd --device /dev/dri` flags. To enable GPU for Qdrant you need to set enable flag. Example of starting Qdrant with AMD GPU support:

```bash
# `--device /dev/kfd --device /dev/dri` flags say to Docker that we want to use GPUs.
# `-e QDRANT__GPU__indexing=1` flag says to QDrant that we want to use GPUs for indexing.
docker run \
	--rm \
	--device /dev/kfd --device /dev/dri \
	-e QDRANT__log_level=debug \
	-p 6333:6333 \
	-p 6334:6334 \
	-e QDRANT__GPU__indexing=1 \
	qdrant/qdrant:gpu-amd-latest
```

Check logs to confirm GPU initialization. Example log output:

```text
2025-01-10T11:56:55.926466Z  INFO gpu::instance: Foung GPU device: AMD Radeon Graphics (RADV GFX1103_R1)
2025-01-10T11:56:55.926485Z  INFO gpu::instance: Foung GPU device: llvmpipe (LLVM 17.0.6, 256 bits) 
2025-01-10T11:56:55.926504Z  INFO gpu::device: Create GPU device AMD Radeon Graphics (RADV GFX1103_R1)
```

This concludes the setup. In a basic scenario, you won't need to configure anything else.

## Known limitations

> * **Platform Support:** Only Linux x86_64 is supported. Windows, macOS, ARM, and other platforms are not supported.

Reviewer 2: Linux x64 is a limitation of docker image. Is user run from binary, it doesn't matter. I'm working under windows as well
